src/watertap_contrib/seto/unit_models/zero_order/chemical_softening_0D.py

```diff
@@ -0,0 +1,949 @@
+from copy import deepcopy
+
+# check
+# Import Pyomo libraries
+from pyomo.environ import (
+    Set,
+    Var,
+    Constraint,
+    check_optimal_termination,
+    Param,
+    Suffix,
+    value,
+    log,
+    units as pyunits,
+    Expression,
+    NonNegativeReals,
+)
+from pyomo.common.config import ConfigBlock, ConfigValue, In
+
+# Import IDAES cores
+from idaes.core import (
+    declare_process_block_class,
+    UnitModelBlockData,
+    useDefault,
+)
+from idaes.core.solvers.get_solver import get_solver
+from idaes.core.util.tables import create_stream_table_dataframe
+from idaes.core.util.constants import Constants
+from idaes.core.util.config import is_physical_parameter_block
+from idaes.core.util.misc import StrEnum
+
+from idaes.core.util.exceptions import ConfigurationError, InitializationError
+import idaes.core.util.scaling as iscale
+import idaes.logger as idaeslog
+
+from watertap.core import InitializationMixin
+
+__author__ = "Kurban Sitterley, Abdiel Lugo"
+
+_log = idaeslog.getLogger(__name__)
+
+
+class SofteningTrainType(StrEnum):
+    conventional = "conventional"
+    solid_contact = "solid_contact"
+
+
+class SofteningProcedureType(StrEnum):
+    single_stage_lime = "single_stage_lime"
+    excess_lime = "excess_lime"
+    single_stage_lime_soda = "single_stage_lime_soda"
+    excess_lime_soda = "excess_lime_soda"
+
+
+@declare_process_block_class("ChemicalSoftening0D")
```

Comment:
I am not 100% clear. Is this a 0D model or a zero-order (assumed performance across the board) model? I ask because this file is named 0D but is in the zero_order folder. Meanwhile you also have a chem softening 0d unit in the units folder. I haven't reviewed that file yet; if they are different and this is meant to be a simpler ZO representation, please change name to `ChemicalSofteningZO`

Reply:
I agree. I started this model (and am responsible for the name) with aspirations to make it more 0D-ish. But it is definitely a ZO model.